### PR TITLE
Type Hinting and Type Aliasing Draft PR

### PR DESCRIPTION
**Pull request by ABostrom**
#### Reference Issues/PRs
This is a draft PR for type hinting in SKtime. 

It's super WIP, but I wanted to start work on it. so others could see my ideas, and comment as I muddle my way through.

I needed a long summer project I could work on slowly. 

#### What should a reviewer concentrate their feedback on?
Typing the data containers is tricky, as we have some different formats. 

Everything else should be straightforward.

### CODE REVIEW

**Comment by ABostrom** [comment]
If people want to contribute and help:

Things I'd like others to do some research on. Typing / Aliasing nested pandas data frames. Typing numpy arrays.

**Comment by fkiraly** [comment]
So, my big question is: what is a "good" way to do runtime type checks or hints, using "implicit type definitions" such as "a `pandas.DataFrame` with exactly two columns named `Alice` and `Bob` with no more than 100 rows"?

**Review by mloning — COMMENTED** [review]
@ABostrom looks good to me. Just two minor comments.

Is there any way to check if these type hints are consistent with the code in the repo? We could think about adding that to the CI.

We also need to make sure to replace the type hints [here](https://github.com/alan-turing-institute/sktime/blob/main/sktime/transformations/base.py) to avoid having duplicate definitions of types.

File: sktime/utils/_typing.py [hunk 1]
```diff
@@ -0,0 +1,32 @@
+from typing import Union, Tuple
```
[the remark this answers is not in the document]
Comment (ABostrom) [reply sktime/utils/_typing.py:1]: yeah. happy to move it. Just picked somewhere for the time being.

**Comment by ABostrom** [comment]
> 
> 
> So, my big question is: what is a "good" way to do runtime type checks or hints, using "implicit type definitions" such as "a `pandas.DataFrame` with exactly two columns named `Alice` and `Bob` with no more than 100 rows"?

So i think this is a pretty open question within python world.

**Comment by chrisholder** [comment]
Thought Id jump in and just add my take on this. So first off just some links to typing packages I've used in other projects and are extremely useful we may want to consider (I've linked these to @ABostrom in the past) :
https://pypi.org/project/nptyping/
https://pypi.org/project/data-science-types/

So just quickly how I structure my types for projects is:
- Have a file in the base directory called base_types.py which is intended to contain 'shared types'. These are utility esc types for things non project specific like numpy types, dataframe types etc.
- For types specific to a class or function I declare them in the same file they exist in.

My motivation for doing it this way doesn't come from Python. As @ABostrom said above it's quite an open conversation currently in the Python community, so I decided to structure my types like I would as if I was using Typescript. Typescript is similar in the sense it's built on top of a language that doesn't have types (javascript) . Typescript is just a wrapper around your javascript which allows type checking to be done at run time. Once checks are complete the wrapper is removed and plain javascript is outputted. This is essentially doing the same thing python typing is doing (although Typescript is more established and more powerful for typing stuff).

While I certainly don't think this is the final end solution potentially drawing some of how Typescript does types could be of benefit just because it is so well established and used.

**Comment by fkiraly** [comment]
maybe also useful to point out this old enhancement proposal of mine:
https://github.com/sktime/enhancement-proposals/blob/main/steps/05_scitype_based_IO_checks/step.md

look at the "type checking" at the bottom, this could be its own python package.

Regarding the main content, I think we're moving towards the nested architecture (see foreceasting base class) instead of the decorator architecture discussed in the STEP.

**Comment by fkiraly** [comment]
@ABostrom, @chrisholder, do you know anything good for *declarative* typing?

E.g., I want to *assign* a type `"mytype"` to a variable `x`, so I can do sth like first `assign_custom_type(x, "mytype")`, then query `custom_type_of(x)` which should return `"mytype"`? And/or the ability to define your own types and conduct inference on them.

As opposed to *inferring* type based on some pre-defined or pre-existing type system.

**Comment by chrisholder** [comment]
> @ABostrom, @chrisholder, do you know anything good for _declarative_ typing?
> 
> E.g., I want to _assign_ a type `"mytype"` to a variable `x`, so I can do sth like first `assign_custom_type(x, "mytype")`, then query `custom_type_of(x)` which should return `"mytype"`? And/or the ability to define your own types and conduct inference on them.
> 
> As opposed to _inferring_ type based on some pre-defined or pre-existing type system.

Not 100% sure what you mean but did this example of what I thought you mean. You can't guarentee that someone will assign the correct value but you can definitely add additional checks.

image

I also saw this in the python documentation:

image

**Comment by fkiraly** [comment]
@chrisholder, see #980 and the linked extension proposal no.5 therein - hope that makes it clearer

**Comment by ABostrom** [comment]
> https://pypi.org/project/data-science-types/

This has been depcreated and abandoned, based on pandas and numpy pushing this internally.

**Comment by ABostrom** [comment]
The only dependency I would like to propose we add is : https://github.com/python/mypy

Mypy can be sued to staticly check the type hints we add to our programs, and I think can be a big push in the direction of code correctness validation.

**Comment by mloning** [comment]
> Mypy can be sued to staticly check the type hints we add to our programs, and I think can be a big push in the direction of code correctness validation.

That sounds like a good idea, having this run in the CI pipelines would be important to ensure that that code/type hints make sense

**Comment by mloning** [comment]
Sorry, accidentally closed the PR, reopened now.

**Comment by TonyBagnall** [comment]
hi @ABostrom , there is conflict with main now, can we close this PR and reopen it from a branch? That way I can tidy it up

**Comment by fkiraly** [comment]
@abostrom, ping

**Comment by mloning** [comment]
I'll close this for now, feel free to re-open if you want to pick this up again, I've also started playing around with it here #1679

**Pull request closed without merging**